File: Hexarc.Pact.Tool.TypeScript/src/utils/client_settings_reader.ts
import * as File from "./file";
import type { ClientSettings } from "../types/tool";


export async function read(): Promise<ClientSettings[] | null> {
  if (!await File.exists("pact.json")) return null;

  const raw = await File.read("pact.json");
  const document = JSON.parse(raw);

  if (Array.isArray(document)) return document;
  else return [document];
}

File: Hexarc.Pact.Tool/Internals/ClientSettingsReader.cs
using Hexarc.Pact.Tool.Models;

namespace Hexarc.Pact.Tool.Internals;

public sealed class ClientSettingsReader
{
    private const String SettingsFileName = "pact.json";

    private JsonSerializerOptions JsonSerializerOptions { get; }

    public ClientSettingsReader(JsonSerializerOptions jsonSerializerOptions) =>
        this.JsonSerializerOptions = jsonSerializerOptions;

    public async Task<IEnumerable<ClientSettings>?> Read()
    {
        if (!File.Exists(SettingsFileName)) return default;

        await using var stream = File.OpenRead(SettingsFileName);
        using var document = await JsonDocument.ParseAsync(stream);

        return document.RootElement.ValueKind switch
        {
            JsonValueKind.Array => document.Deserialize<ClientSettings[]>(this.JsonSerializerOptions),
            _ => new[] { document.Deserialize<ClientSettings>(this.JsonSerializerOptions) ?? throw new NullReferenceException() }
        };
    }
}
